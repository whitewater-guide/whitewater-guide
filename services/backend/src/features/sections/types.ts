import { RawTimestamped } from '@db';
import { PointRaw } from '@features/points';
import {
  Coordinate3d,
  NamedNode,
  SectionEditAction,
  Tag,
} from '@whitewater-guide/commons';

export interface GaugeBindingRaw {
  minimum?: number;
  maximum?: number;
  optimum?: number;
  impossible?: number;
  approximate?: boolean;
}

interface ShapeRaw {
  type: string;
  coordinates: Coordinate3d[];
}

export interface SectionRaw extends NamedNode, RawTimestamped {
  alt_names: string[];

  river_id: string;
  river_name: string; // for embedding
  gauge_id: string | null;
  region_id: string;
  region_name: string; // for embedding

  hidden: boolean;
  demo: boolean;
  premium: boolean;

  season_numeric: number[];
  levels: GaugeBindingRaw | null;
  flows: GaugeBindingRaw | null;
  shape: ShapeRaw;
  distance: number | null;
  drop: number | null;
  duration: number | null;
  difficulty: number;
  difficulty_xtra: string | null;
  rating: number | null;
  description: string | null;
  season: string | null;
  flows_text: string | null;
  put_in: string;
  take_out: string;

  pois: PointRaw[] | null;
  tags: Tag[] | null;

  count: number | null; // window function count
}

export interface SectionsFilter {
  riverId?: string;
  regionId?: string;
  updatedAfter?: Date;
}

export interface SectionsEditLogRaw {
  id?: string; // optional for input
  section_id: string;
  old_section_name: string;
  new_section_name: string;
  river_id: string;
  river_name: string;
  region_id: string;
  region_name: string;
  editor_id: string;
  editor_name: string;
  action: SectionEditAction;
  created_at: Date;
}
