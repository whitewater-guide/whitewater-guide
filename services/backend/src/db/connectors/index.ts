export * from './types';
export * from './OffsetConnector';
export * from './createConnectors';
