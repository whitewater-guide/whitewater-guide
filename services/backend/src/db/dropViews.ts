import Knex from 'knex';

export async function dropViews(db: Knex, ...views: string[]) {
  for (const view of views) {
    await db.raw(`DROP VIEW IF EXISTS ${view}_view`);
  }
}
