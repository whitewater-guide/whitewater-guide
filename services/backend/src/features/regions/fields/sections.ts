import { Context, ListQuery } from '~/apollo';
import { SectionsFilter } from '~/features/sections';
import { GraphQLFieldResolver } from 'graphql';
import { RegionRaw } from '../types';

interface Vars extends ListQuery {
  filter?: SectionsFilter;
}

const sectionsResolver: GraphQLFieldResolver<RegionRaw, Context, Vars> = async (
  { id },
  { page, filter = {} },
  { dataSources },
  info,
) => {
  const { updatedAfter } = filter;
  return dataSources.sections.getMany(info, {
    page,
    filter: { regionId: id, updatedAfter },
  });
};

export default sectionsResolver;
