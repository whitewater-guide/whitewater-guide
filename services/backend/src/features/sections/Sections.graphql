type GaugeBinding {
  # Below this level the river is not enjoyable to paddle
  minimum: Float
  # The river is still runnable above this level, but difficulty is considerably higher than mentioned here
  maximum: Float
  # The best level for paddlers who consider themselves to be in the same class as the section
  optimum: Float
  # It is insanely dangerous or physically impossible to run river this high
  impossible: Float
  # If true, then the gauge gives very vague idea of actual section flows (e.g. the gauge is on other tributary)
  approximate: Boolean
  # Optional formula to compute section flow from gauge value, e.g '(x + 10) * 2' where x is gauge value
  formula: String
}

input GaugeBindingInput {
  minimum: Float
  maximum: Float
  optimum: Float
  impossible: Float
  approximate: Boolean
  formula: String
}

type Section {
  # Unique section ID
  id: ID!
  # Human-friendly section name (without river name, e.g. just 'Upper')
  name: String!
  # Alternative names, e.g. english transliteration for names that contain umlauts
  altNames: [String!]
  # Guidebook page. Subject to premium access limitations
  # Returns null if requires premium access
  # Returns empty string if no description provided
  description: String
  # Human-friendly description to complement numeric months list
  season: String
  # Array of numers between 0 and 23, corresponding to half-months
  seasonNumeric: [Int]!

  # Section region
  region: Region
  # Section river
  river: River

  # Section gauge
  gauge: Gauge
  # Level bindings bindings
  levels: GaugeBinding
  # Flow gauge bindings
  flows: GaugeBinding
  # Human-friendly description to complement gauge bindings, e.g. 'After heavy rainfall'
  flowsText: String

  # Put-in location
  putIn: Point!
  # Take-out location
  takeOut: Point!
  # Array of [lon, lat, alt] points to roughly match river shape, ordered (inclusively) from put-in to take-out
  shape: [[Float!]]!
  # Approximate section length, in km
  distance: Float
  # Approximate difference between put-in and take-out altitudes, in m
  drop: Float
  # Dictionay value, approximate time required to run this section
  duration: Int
  # Floating point difficulty value, i.e. 4.5 means IV+
  difficulty: Float!
  # Short extra difficulty to be written in brackets, i.e. what's inside brackets for "IV (2xV, X)" difficuly
  difficultyXtra: String
  # Subjective rating from 0 to 5 stars, half stars allowed
  rating: Float
  # Array of tags
  tags: [Tag]!
  # When was this section added to whitewater.guide
  createdAt: Date
  # When was this section last modified
  updatedAt: Date
  # Photos, videos and blogs
  media(page: Page): SectionMediaConnection!
  # Points of interest
  pois: [Point]!

  # Not ready for publication yet
  hidden: Boolean!
  # Is demo section in premium region
  demo: Boolean
}

type SectionMediaConnection {
  nodes: [Media]!
  count: Int!
}

input SectionInput {
  id: ID
  name: String!
  altNames: [String!]
  description: String
  season: String
  seasonNumeric: [Int]

  river: RefInput!
  gauge: RefInput
  # Optional, but required when creating new river
  region: RefInput
  levels: GaugeBindingInput
  flows: GaugeBindingInput
  flowsText: String

  shape: [[Float!]]!
  distance: Float
  drop: Float
  duration: Int
  difficulty: Float!
  difficultyXtra: String
  rating: Float
  tags: [RefInput]!
  pois: [PointInput]!

  hidden: Boolean!
}

input SectionAdminSettings {
  demo: Boolean!
}

type SectionsList {
  nodes: [Section]!
  count: Int!
}

input SectionsFilter {
  riverId: ID
  regionId: ID
  updatedAfter: Date
}

type SectionEditLogEntry {
  id: ID!
  section: Section!
  editor: User!
  action: String!
  diff: JSON
  createdAt: Date!
}

input SectionsEditLogFilter {
  regionId: ID
  editorId: ID
}

type SectionsEditLogList {
  nodes: [SectionEditLogEntry]!
  count: Int!
}

type Query {
  sections(page: Page, filter: SectionsFilter): SectionsList!
  section(id: ID): Section
  sectionsEditLog(
    filter: SectionsEditLogFilter
    page: Page
  ): SectionsEditLogList! @admin
}

type Mutation {
  upsertSection(section: SectionInput!): Section
  removeSection(id: ID!): ID
  administrateSection(id: ID!, settings: SectionAdminSettings!): Section @admin
}
