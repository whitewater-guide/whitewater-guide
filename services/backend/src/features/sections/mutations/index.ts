import administrateSection from './administrateSection';
import bulkInsert from './bulkInsert';
import removeSection from './removeSection';
import upsertSection from './upsertSection';

export default {
  administrateSection,
  bulkInsert,
  removeSection,
  upsertSection,
};
