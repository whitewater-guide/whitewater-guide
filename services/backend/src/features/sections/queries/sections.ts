import { ListQuery, TopLevelResolver } from '@apollo';
import { SectionsFilter } from '../types';

interface Vars extends ListQuery {
  filter?: SectionsFilter;
}

const sections: TopLevelResolver<Vars> = async (
  _,
  { filter, page },
  { dataSources },
  info,
) => {
  const result = await dataSources.sections.getMany(info, { filter, page });
  return result;
};

export default sections;
