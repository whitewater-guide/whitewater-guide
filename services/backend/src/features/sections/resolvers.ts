import { listResolvers } from '@apollo';
import { gaugeBindingResolver, sectionFieldResolvers } from './fields';
import Mutation from './mutations';
import Query from './queries';

export const sectionsResolvers = {
  Section: sectionFieldResolvers,
  GaugeBinding: gaugeBindingResolver,
  SectionsList: listResolvers,
  SectionMediaConnection: listResolvers,
  Query,
  Mutation,
};
