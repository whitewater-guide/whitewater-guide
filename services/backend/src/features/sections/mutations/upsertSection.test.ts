import db, { holdTransaction, rollbackTransaction } from '@db';
import { SectionsEditLogRaw } from '@features/sections';
import {
  fileExistsInBucket,
  MEDIA,
  MEDIA_BUCKET_DIR,
  resetTestMinio,
  TEMP,
  TEMP_BUCKET_DIR,
} from '@minio';
import {
  ADMIN,
  ADMIN_ID,
  EDITOR_GA_EC,
  EDITOR_GA_EC_ID,
  EDITOR_NO_EC,
  EDITOR_NO_EC_ID,
  TEST_USER,
} from '@seeds/01_users';
import { REGION_GALICIA, REGION_NORWAY } from '@seeds/04_regions';
import { RIVER_GAL_1, RIVER_SJOA } from '@seeds/07_rivers';
import { GALICIA_R1_S1, NORWAY_SJOA_AMOT } from '@seeds/09_sections';
import { PHOTO_1 } from '@seeds/11_media';
import { MEDIA_SUGGESTION_ID1 } from '@seeds/17_suggestions';
import {
  countRows,
  fakeContext,
  noUnstable,
  runQuery,
  UUID_REGEX,
} from '@test';
import {
  Duration,
  MediaInput,
  MediaKind,
  NEW_ID,
  SectionInput,
} from '@whitewater-guide/commons';
import { copy } from 'fs-extra';
import { ExecutionResult } from 'graphql';
import set from 'lodash/fp/set';
import * as path from 'path';

let spBefore: number;
let pBefore: number;
let rBefore: number;
let sBefore: number;
let tBefore: number;
let suggBefore: number;

beforeAll(async () => {
  [spBefore, pBefore, rBefore, sBefore, tBefore, suggBefore] = await countRows(
    true,
    'sections_points',
    'points',
    'rivers',
    'sections',
    'sections_tags',
    'suggested_sections',
  );
});

beforeEach(async () => {
  await holdTransaction();
  await resetTestMinio();
});
afterEach(rollbackTransaction);
afterAll(() => resetTestMinio(true));

const upsertQuery = `
  mutation upsertSection($section: SectionInput!){
    upsertSection(section: $section){
      id
      name
      description
      season
      seasonNumeric
      region {
        id
        name
      }
      river {
        id
        name
      }
      gauge {
        id
        name
      }
      levels {
        minimum
        optimum
        maximum
        impossible
        approximate
      }
      flows {
        minimum
        optimum
        maximum
        impossible
        approximate
      }
      flowsText
      putIn {
        id
        name
        coordinates
      }
      takeOut {
        id
        name
        coordinates
      }
      shape
      distance
      drop
      duration
      difficulty
      difficultyXtra
      rating
      tags {
        id
        name
      }
      createdAt
      updatedAt
      hidden
      helpNeeded
      demo
      pois {
        id
        name
        coordinates
      }
      media {
        nodes {
          id
          kind
          url
        }
        count
      }
    }
  }
`;

const existingRiverSection: SectionInput = {
  id: null,
  name: 'Playrun',
  altNames: null,
  description: 'Playrun description',
  season: 'Playrun season',
  seasonNumeric: [1, 2, 3],
  river: {
    id: RIVER_SJOA,
  },
  gauge: {
    id: 'c03184b4-aaa0-11e7-abc4-cec278b6b50a',
  },
  levels: {
    minimum: 100,
    optimum: null,
    maximum: null,
    impossible: null,
    approximate: false,
  },
  flows: null,
  flowsText: 'Playrun flows',

  shape: [[33, 33], [34, 42, 0], [34, 43, null]],
  distance: 2.44,
  drop: 101.1,
  duration: Duration.LAPS,
  difficulty: 4,
  difficultyXtra: 'X',
  rating: 5,
  tags: [{ id: 'waterfalls' }, { id: 'undercuts' }],
  pois: [
    {
      id: null,
      name: 'playrun pt 1',
      description: 'pt 1 d',
      kind: 'other',
      coordinates: [10, 12, 0], // 3d
    },
    {
      id: null,
      name: 'playrun pt 2',
      description: 'pt 2 d',
      kind: 'portage',
      coordinates: [33, 34], // 2d
    },
  ],
  hidden: false,
  helpNeeded: 'please proofread this',
};

const newRiverSection = {
  ...existingRiverSection,
  river: { id: NEW_ID, name: 'Rauma' },
  region: { id: REGION_NORWAY },
};

const updateData: SectionInput = {
  ...existingRiverSection,
  id: GALICIA_R1_S1, // galician river 1 section 1
  river: {
    id: RIVER_GAL_1,
  },
  gauge: {
    id: 'aba8c106-aaa0-11e7-abc4-cec278b6b50a', // Galicia gauge 1
  },
  // Seed section has 2 POIS
  // Delete 1 POI, update 1 poi, add 2 POIs
  pois: [
    {
      id: 'ca0bee06-d445-11e7-9296-cec278b6b50a',
      name: 'Updated poi name', // was 'Galicia Riv 1 Sec 1 Rapid',
      description: 'Updated poi description ', // was 'Some rapid',
      kind: 'other', // was rapid
      coordinates: [2.1, 2.3, 3.4], // was [1.2, 3.2, 4.3]
    },
    {
      id: null,
      name: 'Updated new poi 1 name',
      description: 'Updated new poi 1 description',
      kind: 'portage',
      coordinates: [66, 77, null],
    },
    {
      id: null,
      name: 'Updated new poi 2 name',
      description: 'Updated new poi 2 description',
      kind: 'other',
      coordinates: [50.1, 50.2, 333],
    },
  ],
  // Delete one tag, add two new tags, keep one old
  tags: [{ id: 'creeking' }, { id: 'undercuts' }, { id: 'snowmelt' }],
  helpNeeded: 'fix take-out',
};

const invalidSection: SectionInput = {
  id: null,
  name: 'z',
  altNames: ['x'],
  description: null,
  season: null,
  seasonNumeric: [300, 2, 3],
  river: {
    id: RIVER_SJOA,
  },
  gauge: null,
  levels: null,
  flows: null,
  flowsText: null,

  shape: [[300, 33, 0], [34, 42, 0]],
  distance: -2.44,
  drop: -101.1,
  duration: 4,
  difficulty: 33,
  difficultyXtra: null,
  rating: 35,
  tags: [],
  pois: [],

  hidden: false,
  helpNeeded: null,
};

it('should fail on invalid input', async () => {
  const result = await runQuery(
    upsertQuery,
    { section: invalidSection },
    fakeContext(ADMIN),
  );
  expect(result).toHaveGraphqlValidationError();
});

it.each([
  ['anon', undefined],
  ['user', TEST_USER],
  ['non-owning editor', EDITOR_GA_EC],
])('%s should create suggestion', async (_, user: any) => {
  const result = await runQuery(
    upsertQuery,
    { section: newRiverSection },
    fakeContext(user),
  );
  expect(result.errors).toBeUndefined();
  const [spAfter, pAfter, rAfter, sAfter, tAfter, suggAfter] = await countRows(
    false,
    'sections_points',
    'points',
    'rivers',
    'sections',
    'sections_tags',
    'suggested_sections',
  );
  expect([
    spAfter - spBefore,
    pAfter - pBefore,
    rAfter - rBefore,
    sAfter - sBefore,
    tAfter - tBefore,
    suggAfter - suggBefore,
  ]).toEqual([0, 0, 0, 0, 0, 1]);
});

describe('insert', () => {
  let insertResult: any;
  let insertedSection: any;

  beforeEach(async () => {
    insertResult = await runQuery(
      upsertQuery,
      { section: existingRiverSection },
      fakeContext(EDITOR_NO_EC),
    );
    insertedSection =
      insertResult && insertResult.data && insertResult.data.upsertSection;
  });

  afterEach(() => {
    insertResult = null;
    insertedSection = null;
  });

  it('should return result', () => {
    expect(insertResult.errors).toBeUndefined();
    expect(noUnstable(insertedSection)).toMatchSnapshot();
  });

  it('should insert into tables', async () => {
    const [sections] = await countRows(false, 'sections');
    expect(sections - sBefore).toBe(1);
  });

  it('should not change total number of rivers', async () => {
    const [rivers] = await countRows(false, 'rivers');
    expect(rivers - rBefore).toBe(0);
  });

  it('should insert pois', async () => {
    const [spAfter, pAfter] = await countRows(
      false,
      'sections_points',
      'points',
    );
    expect([spAfter - spBefore, pAfter - pBefore]).toEqual([2, 2]);
  });

  it('should insert tags', async () => {
    const [tags] = await countRows(false, 'sections_tags');
    expect(tags - tBefore).toBe(2);
  });

  it('should correctly set created_by', async () => {
    const { created_by } = await db()
      .table('sections')
      .select(['created_by'])
      .where({ id: insertedSection.id })
      .first();
    expect(created_by).toBe(EDITOR_NO_EC_ID);
  });

  it('should log this event', async () => {
    const entry: SectionsEditLogRaw = await db(false)
      .table('sections_edit_log')
      .orderBy('created_at', 'desc')
      .select('*')
      .first();
    expect(entry).toMatchObject({
      id: expect.stringMatching(UUID_REGEX),
      section_id: insertedSection.id,
      section_name: existingRiverSection.name,
      river_id: existingRiverSection.river.id,
      river_name: 'Sjoa',
      region_id: REGION_NORWAY,
      region_name: 'Norway',
      editor_id: EDITOR_NO_EC_ID,
      action: 'create',
      diff: null,
      created_at: expect.any(Date),
    });
  });
});

describe('insert with new river', () => {
  let insertResult: any;
  let insertedSection: any;

  beforeEach(async () => {
    insertResult = await runQuery(
      upsertQuery,
      {
        section: newRiverSection,
      },
      fakeContext(EDITOR_NO_EC),
    );
    insertedSection =
      insertResult && insertResult.data && insertResult.data.upsertSection;
  });

  afterEach(() => {
    insertResult = null;
    insertedSection = null;
  });

  it('should return result', () => {
    expect(insertResult.errors).toBeUndefined();
    expect(insertedSection).toMatchObject({
      river: {
        id: expect.stringMatching(UUID_REGEX),
        name: 'Rauma',
      },
    });
  });

  it('should insert into tables', async () => {
    const [sections, rivers] = await countRows(false, 'sections', 'rivers');
    expect(sections - sBefore).toBe(1);
    expect(rivers - rBefore).toBe(1);
  });
});

describe('update', () => {
  let updateResult: any;
  let updatedSection: any;
  let originalSection: any;
  let oldPoiIds: any;

  beforeAll(async () => {
    originalSection = await db(true)
      .table('sections_view')
      .where({ id: updateData.id })
      .first();
    oldPoiIds = originalSection.pois.map((poi: any) => poi.id);
    expect(oldPoiIds).toHaveLength(2);
  });

  beforeEach(async () => {
    updateResult = await runQuery(
      upsertQuery,
      { section: updateData },
      fakeContext(EDITOR_GA_EC),
    );
    updatedSection =
      updateResult && updateResult.data && updateResult.data.upsertSection;
  });

  afterEach(() => {
    updateResult = null;
    updatedSection = null;
  });

  it('should return result', async () => {
    expect(updateResult.errors).toBeUndefined();
    expect(updatedSection.id).toBe(updateData.id);
    expect(noUnstable(updatedSection)).toMatchSnapshot();
  });

  it('should not change total number of sections', async () => {
    const [sections] = await countRows(false, 'sections');
    expect(sections - sBefore).toBe(0);
  });

  it('should not change total number of rivers', async () => {
    const [rivers] = await countRows(false, 'rivers');
    expect(rivers - rBefore).toBe(0);
  });

  it('should increase updated_at timestamp', () => {
    expect(updatedSection.createdAt).toBe(
      originalSection!.created_at.toISOString(),
    );
    expect(new Date(updatedSection.updatedAt).valueOf()).toBeGreaterThan(
      originalSection!.updated_at.valueOf(),
    );
  });

  it('should change the number of pois', async () => {
    const [spAfter, pAfter] = await countRows(
      false,
      'sections_points',
      'points',
    );
    expect([spAfter - spBefore, pAfter - pBefore]).toEqual([1, 1]);
  });

  it('should delete POI', async () => {
    const sp = await db()
      .table('sections_points')
      .whereIn('point_id', oldPoiIds)
      .count()
      .first();
    expect(Number(sp.count)).toBe(1);
    const p = await db()
      .table('points')
      .whereIn('id', oldPoiIds)
      .count()
      .first();
    expect(Number(p.count)).toBe(1);
  });

  it('should insert pois', async () => {
    const sp = await db()
      .table('sections_points')
      .where({ section_id: updateData.id })
      .whereNotIn('point_id', oldPoiIds)
      .count()
      .first();
    expect(Number(sp.count)).toBe(2);
  });

  it('should update poi', async () => {
    const point = await db()
      .table('points_view')
      .select('name')
      .where({ id: 'ca0bee06-d445-11e7-9296-cec278b6b50a', language: 'en' })
      .first();
    expect(point.name).toBe('Updated poi name');
  });

  it('should change number of tags', async () => {
    const [tags] = await countRows(false, 'sections_tags');
    expect(tags - tBefore).toBe(1);
  });

  it('should not modify created_by', async () => {
    const { created_by } = await db()
      .table('sections')
      .select(['created_by'])
      .where({ id: updatedSection.id })
      .first();
    expect(created_by).toBe(ADMIN_ID);
  });

  it('should log this event', async () => {
    const entry: SectionsEditLogRaw = await db(false)
      .table('sections_edit_log')
      .orderBy('created_at', 'desc')
      .select('*')
      .first();
    expect(entry).toMatchObject({
      id: expect.stringMatching(UUID_REGEX),
      section_id: updateData.id,
      section_name: updateData.name,
      river_id: updateData.river.id,
      river_name: 'Gal_Riv_One',
      region_id: REGION_GALICIA,
      region_name: 'Galicia',
      editor_id: EDITOR_GA_EC_ID,
      action: 'update',
      diff: expect.any(Object),
      created_at: expect.any(Date),
    });
  });
});

describe('i18n', () => {
  const amotFr: SectionInput = {
    id: NORWAY_SJOA_AMOT,
    name: 'Amot FR',
    altNames: ['Le Amot'],
    river: {
      id: RIVER_SJOA,
    },
    seasonNumeric: [10, 11, 12, 13, 14, 15, 16],
    shape: [[1, 1, 0], [2, 2, 0]],
    distance: 3.2,
    drop: 34.2,
    duration: Duration.LAPS,
    difficulty: 4,
    rating: 1,
    description: 'Amot description FR',
    season: 'Amot season FR',
    flowsText: 'Amot flows FR',
    difficultyXtra: null,
    flows: null,
    pois: [],
    gauge: null,
    levels: null,
    tags: [],

    hidden: false,
    helpNeeded: null,
  };

  it('should add translation', async () => {
    const upsertResult = await runQuery(
      upsertQuery,
      { section: amotFr },
      fakeContext(EDITOR_NO_EC, 'fr'),
    );
    expect(upsertResult.errors).toBeUndefined();
    const translation = await db()
      .table('sections_translations')
      .select()
      .where({ section_id: amotFr.id, language: 'fr' })
      .first();
    expect(translation.name).toBe('Amot FR');
  });

  it('should modify common props when translation is added', async () => {
    const upsertResult = await runQuery(
      upsertQuery,
      { section: amotFr },
      fakeContext(EDITOR_NO_EC, 'fr'),
    );
    expect(upsertResult.errors).toBeUndefined();
    const section = await db()
      .table('sections')
      .select()
      .where({ id: amotFr.id })
      .first();
    expect(section.rating).toBe(1);
  });

  it('should modify translation', async () => {
    const upsertResult = await runQuery(
      upsertQuery,
      { section: amotFr },
      fakeContext(EDITOR_NO_EC, 'ru'),
    );
    expect(upsertResult.errors).toBeUndefined();
    const translation = await db()
      .table('sections_translations')
      .select()
      .where({ section_id: amotFr.id, language: 'ru' })
      .first();
    expect(translation).toEqual(
      expect.objectContaining({
        description: 'Amot description FR',
        season: 'Amot season FR',
        flows_text: 'Amot flows FR',
      }),
    );
  });

  it('should update poi translation', async () => {
    const updateRu = {
      ...updateData,
      pois: [
        {
          id: 'ca0bee06-d445-11e7-9296-cec278b6b50a',
          name: 'Русской имя', // was 'Galicia Riv 1 Sec 1 Rapid',
          description: 'Русское описание ', // was 'Some rapid',
          kind: 'other', // was rapid
          coordinates: [2.1, 2.3, 3.4], // was [1.2, 3.2, 4.3]
        },
      ],
    };
    const upsertResult = await runQuery(
      upsertQuery,
      { section: updateRu },
      fakeContext(EDITOR_GA_EC, 'ru'),
    );
    expect(upsertResult.errors).toBeUndefined();
    const translation = await db()
      .table('points_translations')
      .select()
      .where({ language: 'ru', point_id: updateRu.pois[0].id })
      .first();
    expect(translation).toEqual(
      expect.objectContaining({
        name: 'Русской имя',
        description: 'Русское описание ',
      }),
    );
  });
});

describe('media', () => {
  const media: MediaInput[] = [
    {
      id: null,
      description: 'photo description',
      copyright: 'photo copyright',
      url: 'photo.jpg',
      kind: MediaKind.photo,
      resolution: [100, 100],
      weight: null,
    },
    {
      id: null,
      description: 'video description',
      copyright: 'video copyright',
      url: 'https://www.youtube.com/watch?v=FRmz0QQXHEw',
      kind: MediaKind.video,
      resolution: null,
      weight: null,
    },
  ];
  const insertSection: SectionInput = {
    ...existingRiverSection,
    media,
    suggestionId: MEDIA_SUGGESTION_ID1,
  };
  const updateSection: SectionInput = {
    ...existingRiverSection,
    id: NORWAY_SJOA_AMOT,
    media,
  };

  beforeEach(async () => {
    await copy(
      path.resolve(__dirname, '__tests__/test.jpg'),
      path.resolve(TEMP_BUCKET_DIR, 'photo.jpg'),
    );
    await copy(
      path.resolve(__dirname, '__tests__/test.jpg'),
      path.resolve(MEDIA_BUCKET_DIR, 'media_suggestion1.jpg'),
    );
  });

  describe('insert', () => {
    let result!: ExecutionResult;

    beforeEach(async () => {
      result = await runQuery(
        upsertQuery,
        { section: insertSection },
        fakeContext(EDITOR_NO_EC),
      );
      expect(result.errors).toBeUndefined();
    });

    it('should return media', () => {
      expect(result.data!.upsertSection.media).toMatchObject({
        count: 2,
        nodes: [
          {
            id: expect.stringMatching(UUID_REGEX),
            url: 'photo.jpg',
            kind: MediaKind.photo,
          },
          {
            id: expect.stringMatching(UUID_REGEX),
            url: 'https://www.youtube.com/watch?v=FRmz0QQXHEw',
            kind: MediaKind.video,
          },
        ],
      });
    });

    it('should move photos from temp dir', async () => {
      await expect(fileExistsInBucket(TEMP, 'test.jpg')).resolves.toBe(false);
      await expect(fileExistsInBucket(TEMP, 'photo.jpg')).resolves.toBe(false);
      // expect items bucket to contain new_image.jpg
      await expect(
        fileExistsInBucket(
          MEDIA,
          'photo.jpg',
          '11ad66ff24bc4d463dfe8219f6ba6bcd',
        ),
      ).resolves.toBe(true);
    });
  });

  describe('update', () => {
    let result!: ExecutionResult;

    beforeEach(async () => {
      result = await runQuery(
        upsertQuery,
        { section: updateSection },
        fakeContext(EDITOR_NO_EC),
      );
      expect(result.errors).toBeUndefined();
    });

    it('should return media', () => {
      expect(result.data!.upsertSection.media).toMatchObject({
        count: 2,
        nodes: [
          {
            id: expect.stringMatching(UUID_REGEX),
            url: 'photo.jpg',
            kind: MediaKind.photo,
          },
          {
            id: expect.stringMatching(UUID_REGEX),
            url: 'https://www.youtube.com/watch?v=FRmz0QQXHEw',
            kind: MediaKind.video,
          },
        ],
      });
    });

    it('update should delete existing media', async () => {
      const old = await db(false)
        .select('*')
        .from('media')
        .where({ id: PHOTO_1 });
      expect(old).toEqual([]);
    });

    it('should delete unused media files', async () => {
      await expect(fileExistsInBucket(MEDIA, PHOTO_1)).resolves.toBe(false);
    });
  });
});

it('should sanitize input', async () => {
  let dirty = { ...existingRiverSection, name: "it's a \\ $1 slash with . ?" };
  dirty = set('pois.0.name', "it's a \\ $1 slash with . ?", dirty);
  const result = await runQuery(
    upsertQuery,
    { section: dirty },
    fakeContext(ADMIN),
  );
  expect(result).toHaveProperty(
    'data.upsertSection.name',
    "it's a \\ $1 slash with . ?",
  );
  expect(result).toHaveProperty(
    'data.upsertSection.pois.0.name',
    "it's a \\ $1 slash with . ?",
  );
});
