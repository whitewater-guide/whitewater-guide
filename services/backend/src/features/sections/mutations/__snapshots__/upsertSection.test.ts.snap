// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`insert should return result 1`] = `
Object {
  "demo": false,
  "description": "Playrun description",
  "difficulty": 4,
  "difficultyXtra": "X",
  "distance": 2.44,
  "drop": 101.1,
  "duration": 10,
  "flows": null,
  "flowsText": "Playrun flows",
  "gauge": Object {
    "name": "Norway gauge 1",
  },
  "helpNeeded": "please proofread this",
  "hidden": false,
  "levels": Object {
    "approximate": false,
    "impossible": null,
    "maximum": null,
    "minimum": 100,
    "optimum": null,
  },
  "media": Object {
    "count": 0,
    "nodes": Array [],
  },
  "name": "Playrun",
  "pois": Array [
    Object {
      "coordinates": Array [
        10,
        12,
        0,
      ],
      "name": "playrun pt 1",
    },
    Object {
      "coordinates": Array [
        33,
        34,
        0,
      ],
      "name": "playrun pt 2",
    },
  ],
  "putIn": Object {
    "coordinates": Array [
      33,
      33,
      0,
    ],
    "name": "Sjoa - Playrun: Put-in",
  },
  "rating": 5,
  "region": Object {
    "name": "Norway",
  },
  "river": Object {
    "name": "Sjoa",
  },
  "season": "Playrun season",
  "seasonNumeric": Array [
    1,
    2,
    3,
  ],
  "shape": Array [
    Array [
      33,
      33,
      0,
    ],
    Array [
      34,
      42,
      0,
    ],
    Array [
      34,
      43,
      0,
    ],
  ],
  "tags": Array [
    Object {
      "name": "Waterfalls",
    },
    Object {
      "name": "Undercuts",
    },
  ],
  "takeOut": Object {
    "coordinates": Array [
      34,
      43,
      0,
    ],
    "name": "Sjoa - Playrun: Take-out",
  },
}
`;

exports[`update should return result 1`] = `
Object {
  "demo": false,
  "description": "Playrun description",
  "difficulty": 4,
  "difficultyXtra": "X",
  "distance": 2.44,
  "drop": 101.1,
  "duration": 10,
  "flows": null,
  "flowsText": "Playrun flows",
  "gauge": Object {
    "name": "Galicia gauge 1",
  },
  "helpNeeded": "fix take-out",
  "hidden": false,
  "levels": Object {
    "approximate": false,
    "impossible": null,
    "maximum": null,
    "minimum": 100,
    "optimum": null,
  },
  "media": Object {
    "count": 1,
    "nodes": Array [
      Object {
        "kind": "video",
        "url": "http://some2.video",
      },
    ],
  },
  "name": "Playrun",
  "pois": Array [
    Object {
      "coordinates": Array [
        2.1,
        2.3,
        3.4,
      ],
      "name": "Updated poi name",
    },
    Object {
      "coordinates": Array [
        66,
        77,
        0,
      ],
      "name": "Updated new poi 1 name",
    },
    Object {
      "coordinates": Array [
        50.1,
        50.2,
        333,
      ],
      "name": "Updated new poi 2 name",
    },
  ],
  "putIn": Object {
    "coordinates": Array [
      33,
      33,
      0,
    ],
    "name": "Beca - Playrun: Put-in",
  },
  "rating": 5,
  "region": Object {
    "name": "Galicia",
  },
  "river": Object {
    "name": "Beca",
  },
  "season": "Playrun season",
  "seasonNumeric": Array [
    1,
    2,
    3,
  ],
  "shape": Array [
    Array [
      33,
      33,
      0,
    ],
    Array [
      34,
      42,
      0,
    ],
    Array [
      34,
      43,
      0,
    ],
  ],
  "tags": Array [
    Object {
      "name": "Creeking",
    },
    Object {
      "name": "Undercuts",
    },
    Object {
      "name": "Snowmelt",
    },
  ],
  "takeOut": Object {
    "coordinates": Array [
      34,
      43,
      0,
    ],
    "name": "Beca - Playrun: Take-out",
  },
}
`;
